docs/tasks/manage-gpus/scheduling-gpus.md

```diff
@@ -4,154 +4,203 @@ approvers:
 
+You can consume these GPUs from your containers by requesting for
+`nvidia.com/gpu` just like you request `cpu` or `memory`.
+However, there are some limitations in how you specify the resource requirements
+when using GPUs:
 - GPUs are only supposed to be specified in the `limits` section, which means:
-  * You can specify GPU `limits` without specifying `requests` because Kubernetes
-    will use the limit as the request value by default.
-  * You can specify GPU in both `limits` and `requests` but these two values must equal.
+  * You can specify GPU `limits` without specifying `requests` because
+    Kubernetes will use the limit as the request value by default.
+  * You can specify GPU in both `limits` and `requests` but these two values
+    must equal.
   * You cannot specify GPU `requests` without specifying `limits`.
-- Containers (and pods) do not share GPUs.
-- Each container can request one or more GPUs.
-- It is not possible to request a portion of a GPU.
-- Nodes are expected to be homogenous, i.e. run the same GPU hardware.
-
-If your nodes are running different versions of GPUs, then use Node Labels and Node Selectors to schedule pods to appropriate GPUs.
-Following is an illustration of this workflow:
+- Containers (and pods) do not share GPUs. There's no overcommitting of GPUs.
+- Each container can request one or more GPUs. It is not possible to request a
+  fraction of a GPU.
 
-As part of your Node bootstrapping, identify the GPU hardware type on your nodes and expose it as a node label.
+Unlike with `alpha.kubernetes.io/nvidia-gpu`, when using `nvidia.com/gpu` as
+the resource, you don't have to mount any special directories in your pod
+specs. The device plugin is expected to inject them automatically in the
+container.
 
-```shell
-NVIDIA_GPU_NAME=$(nvidia-smi --query-gpu=gpu_name --format=csv,noheader --id=0 | sed -e 's/ /-/g')
-source /etc/default/kubelet
-KUBELET_OPTS="$KUBELET_OPTS --node-labels='alpha.kubernetes.io/nvidia-gpu-name=$NVIDIA_GPU_NAME'"
-echo "KUBELET_OPTS=$KUBELET_OPTS" > /etc/default/kubelet
-```
-
-Specify the GPU types a pod can use via [Node Affinity](/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) rules.
+Here's an example:
 
 ```yaml
-kind: pod
 apiVersion: v1
+kind: Pod
 metadata:
-  annotations:
-    scheduler.alpha.kubernetes.io/affinity: >
-      {
-        "nodeAffinity": {
-          "requiredDuringSchedulingIgnoredDuringExecution": {
-            "nodeSelectorTerms": [
-              {
-                "matchExpressions": [
-                  {
-                    "key": "alpha.kubernetes.io/nvidia-gpu-name",
-                    "operator": "In",
-                    "values": ["Tesla K80", "Tesla P100"]
-                  }
-                ]
-              }
-            ]
-          }
-        }
-      }
+  name: cuda-vector-add
 spec:
+  restartPolicy: OnFailure
   containers:
-    -
-      name: gpu-container-1
+    - name: cuda-vector-add
```

Comment:
Can we use an officially supported nvidia image ?
Digits, CUDA, CAFFE or DL images such as TF-GPU ?

Reply:
This is the image which is used in Kubernetes e2e tests to test GPU support.

Here's the Dockerfile for it: https://github.com/kubernetes/kubernetes/blob/v1.7.11/test/images/nvidia-cuda/Dockerfile

It's using `nvidia/cuda:8.0-devel-ubuntu16.04` as the base image and other NVIDIA packages.

The reason for using this instead of the base image is that this is a good test of whether things are working or not. If they are not, the pod crash loops.

Added link to Dockerfile from the pod spec.

Reply:
If there's a better image to test GPU support. I am happy to switch to that both here and in the e2e tests.

Reply:
Quoting you :)

> This is describing a task ("how to use GPUs") and not a particular feature.

How about giving access to a GPU tensorflow image in that case.
I don't think that people who want GPU support want to read about the cuda-vector-add image and what it does.

Enabling GPUs is mostly about enabling users to do DL so let's have a DL example

Reply:
> Is not there a limitation with Nvidia offficial images that nvidia-docker is mandatory to use and volume injection will not not be allowed on Nvidia images?

No, we don't support running NVIDIA images without nvidia-docker but we don't actively prevent it.